renderer.js

```diff
@@ -1,160 +1,188 @@
-const baudRate = 9600;
+const baudRate = 115200;
 const maxValue = 1024;
-const serialport = require('serialport');
+const serialport = require("serialport");
+const Readline = require("@serialport/parser-readline");
 let activePort;
+let parser;
 let nSteps = -1;
-let stepDivs = [];
-let canvases = [];
-let textValues = [];
-let textThresholds = [];
-let interval;
+
+let readings = [];
+let thresholds = [];
+
+let intervalHandle;
+
+// HTML elements
+let stepDivElements = [];
+let canvasElements = [];
+let textValueElements = [];
+let textThresholdElements = [];
 
 const init = async () => {
-	await serialport.list().then((ports) => {
-		const options = ports.map((port) => {
-			const option = document.createElement('option');
-			option.appendChild(document.createTextNode(port.path));
-			return option;
-		});
-		const selectedIndex = ports.findIndex((port) =>
-			port.pnpId.startsWith('USB')
-		);
-		if (selectedIndex !== -1) {
-			initializePort(ports[selectedIndex].path);
-			options[selectedIndex].selected = 'selected';
-		}
-		const select = document.querySelector('#ports-dropdown');
-		options.forEach((option) => select.add(option));
-		select.addEventListener('change', (event) => {
-			nSteps = -1;
-			initializePort(event.target.value);
-		});
-	});
+  const ports = await serialport.list();
+
+  const optionElements = ports.map((port) => {
+    const optionElement = document.createElement("option");
+    optionElement.appendChild(document.createTextNode(port.path));
+    return optionElement;
+  });
+
+  const selectedIndex = ports.findIndex((port) => port.pnpId.startsWith("USB"));
+  if (selectedIndex != -1) {
+    initializePort(ports[selectedIndex].path);
+    optionElements[selectedIndex].selected = "selected";
+  }
+
+  const selectElement = document.querySelector("#ports-dropdown");
+  optionElements.forEach((option) => selectElement.add(option));
+  selectElement.addEventListener("change", (event) => {
+    nSteps = -1;
+    initializePort(event.target.value);
+  });
 };
 
 const initializePort = (portName) => {
-	clearInterval(interval);
-	if (activePort) activePort.close();
-	clear();
-	activePort = new serialport(portName, {
-		baudRate: baudRate,
-	});
-	activePort.on('data', (data) => parseData(data));
-	interval = setInterval(queryPort, 10);
+  clearInterval(intervalHandle);
+  if (activePort) activePort.close();
+  clear();
+  activePort = new serialport(portName, {
+    baudRate: baudRate,
+  });
+  parser = activePort.pipe(new Readline({ delimiter: "\n" }));
```

Comment:
This `parser` bit is new. Incoming data goes through `parser` but outgoing data still goes through `activePort`.